src/platform_impl/linux/wayland/window/mod.rs

```diff
@@ -0,0 +1,643 @@
+use std::collections::VecDeque;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::{Arc, Mutex};
+
+use sctk::reexports::client::protocol::wl_surface::WlSurface;
+use sctk::reexports::client::Display;
+
+use sctk::reexports::calloop;
+
+use sctk::window::{
+    ARGBColor, ButtonColorSpec, ColorSpec, ConceptConfig, ConceptFrame, Decorations,
+};
+
+use raw_window_handle::unix::WaylandHandle;
+
+use crate::dpi::{LogicalSize, PhysicalPosition, PhysicalSize, Position, Size};
+use crate::error::{ExternalError, NotSupportedError, OsError as RootOsError};
+use crate::monitor::MonitorHandle as RootMonitorHandle;
+use crate::platform::unix::{ButtonState, Theme};
+use crate::platform_impl::{
+    MonitorHandle as PlatformMonitorHandle, OsError,
+    PlatformSpecificWindowBuilderAttributes as PlatformAttributes,
+};
+use crate::window::{CursorIcon, Fullscreen, WindowAttributes};
+
+use super::env::WindowingFeatures;
+use super::event_loop::WinitState;
+use super::output::{MonitorHandle, OutputManagerHandle};
+use super::{EventLoopWindowTarget, WindowId};
+
+pub mod shim;
+
+use shim::{WindowHandle, WindowRequest, WindowUpdate};
+
+pub struct Window {
+    /// Window id.
+    window_id: WindowId,
+
+    /// The Wayland display.
+    display: Display,
+
+    /// The underlying wl_surface.
+    surface: WlSurface,
+
+    /// The current window size.
+    size: Arc<Mutex<LogicalSize<u32>>>,
+
+    /// A handle to output manager.
+    output_manager_handle: OutputManagerHandle,
+
+    /// Event loop proxy to wake it up.
+    event_loop_awakener: calloop::ping::Ping,
+
+    /// Fullscreen state.
+    fullscreen: Arc<AtomicBool>,
+
+    /// Available windowing features.
+    windowing_features: WindowingFeatures,
+
+    /// Requests that SCTK window should perform.
+    window_requests: Arc<Mutex<Vec<WindowRequest>>>,
+}
+
+impl Window {
+    pub fn new<T>(
+        event_loop_window_target: &EventLoopWindowTarget<T>,
+        attributes: WindowAttributes,
+        platform_attributes: PlatformAttributes,
+    ) -> Result<Self, RootOsError> {
+        let surface = event_loop_window_target
+            .env
+            .create_surface_with_scale_callback(move |scale, surface, mut dispatch_data| {
```

Comment:
>I/O are all events, but for DPI change events with sctk you have to define a callback on the environment, do I get this right?

>That seems very strange.

The thing is that DPI change isn't an event and you should manually track it via OutputListener, since on Wayland that's how you get informed about dpi changes, since surface only has Enter and Leave events with output it moved to. So sctk simulates DPI change event for you by listening to outputs and calling into you when scale was changed. With things like keyboard/pointer/etc the point is that you should destroy and bring them back from time to time. You also ideally want to share the same input listener across those, since downstream application can utilize those. Not even saying that you can bind multiple keyboard/pointer/touch on one seat, so the events for those are multiplexed, however with surface it's a bit different, since those generally represent different surfaces you draw into. Does it make sense?

Reply:
It does make sense, yes. Though the choice still seems rather strange. Now for keyboard/pointer/touch you have it one way and for DPI you do it another. That just seems kinda arbitrary. If sctk provides a helper to subscribe to stuff for you, it feels like it would be more intuitive if that was done for everything or not at all. It's difficult to reason about inconsistent APIs, so finding out how to do something often becomes impossible simply because the methods aren't where you'd expect them to be.

Reply:
>That just seems kinda arbitrary. If sctk provides a helper to subscribe to stuff for you, it feels like it would be more intuitive if that was done for everything or not at all. It's difficult to reason about inconsistent APIs, so finding out how to do something often becomes impossible simply because the methods aren't where you'd expect them to be.

SCTK doesn't provide helpers from pointer and touch events, those are used directly from wayland-client from reexports, since you don't need any additional handling for them. It provides handling for keyboard, since it requires working with libxkbcommon and also repeat rate, so it's quite some code to work. The pointer events that you're getting from ThemedPointer are just events from wayland-client, and it has nothing to do with SCTK.